torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,92 @@
+import torch
+from torch._C import DispatchKey
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+from typing import Any, Optional
+
+"""
+New traceable, functional collectives.
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
```

Comment:
There are two problems with enabling Python dispatcher by default: performance, and torchdeploy/multiply.

Performance: Python dispatcher has a startup cost (when we do the initial resolution; not a big deal) and then a per op overhead (the cost to do a dictionary lookup to see what the Python dispatcher entry should be). It is cheaper than it used to be, but not free. And of course if you actually hit the Python code path in Python dispatcher, this is a sizable slow down (but you can pay this slow down on a per op and dispatch key basis.)

Torchdeploy/multipy: The reason why Python dispatcher works but Python op registration doesn't is because Python dispatcher is TLS and we assume that the Python dispatcher only needs to be enabled whereever we propagate this TLS. If Python dispatcher is on by default, this TLS is mandatory and we MUST propagate it to all threads that do execution on behalf of the Python interpreter. In a generalized multipy setting this implies changes to multipy code, as multipy C++ code that calls into Python interpreters is not required to keep the Python interpreter on the same thread: it can migrate a Python interpreter to another thread and it must know to propagate the TLS in this case. Maybe there is an easy interposition point to actually do this transferral but when I discussed this with @zdevito and @d4l3k I got convinced to not do it this way (and that's why we have hermetic PyObjects instead.)